### PR TITLE
tests: Add test in `cycle_sort.cpp`

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1520"></a>

### CODE REVIEW

**Comment by Swastyy** [comment]
@Panquesito7  Sorry for taking your time in the same kind of file edit/PR again. Its because I contributed for the first time and just to get updates from the master repo to forked repo I forked it again. But I assure you that it won't take long for this PR since you already approved it and I have used the same code as before basically re-created the pull request.

File: sorting/cycle_sort.cpp [hunk 1]
```diff
@@ -1,15 +1,8 @@
 /**
  * @file
- * @brief Implementation of [Cycle
- * sort](https://en.wikipedia.org/wiki/Cycle_sort) algorithm
- *
+ * @brief Implementation of [Cycle sort](https://en.wikipedia.org/wiki/Cycle_sort) algorithm
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:3]: I'd like you to put this back as it was (same in lines 5 and 21).

**Review by Panquesito7 — COMMENTED** [review]
That's the only thing missing! After that, I'll approve this PR. 🙂

File: sorting/cycle_sort.cpp (hunk 1 above)
Comment (Swastyy) [reply sorting/cycle_sort.cpp:3]: They are currently 1-line descriptions.

**Review by mishraabhinn — COMMENTED** [review]
[Reference](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/cycle_sort.cpp)
Do this contribution is some different. We already have a cycle_sort. @Panquesito7

**Comment by Swastyy** [comment]
> [Reference](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/cycle_sort.cpp)
> Do this contribution is some different. We already have a cycle_sort. @Panquesito7



> [Reference](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/cycle_sort.cpp)
> Do this contribution is some different. We already have a cycle_sort. @Panquesito7

https://github.com/TheAlgorithms/C-Plus-Plus/projects/6#card-63413841

File: sorting/cycle_sort.cpp [hunk 2]
```diff
@@ -84,9 +83,8 @@ std::vector<T> cycleSort(const std::vector<T> &in_arr) {
  * @returns void
  */
 static void test() {
-    // [506, 48, 123, 79, 0, 362, 951, 500, 0] return [0, 0, 48, 79, 123, 362,
-    // 500, 506, 951]
-    std::vector<int> array1 = {506, 48, 123, 79, 0, 362, 951, 500, 0};
+    // [4, 3, 2, 1] return [1, 2, 3, 4]
+    std::vector<int> array1 = {4, 3, 2, 1};
```
Comment (mishraabhinn) [inline sorting/cycle_sort.cpp:87]:
```suggestion
    // Test 1
    // [4, 3, 2, 1] return [1, 2, 3, 4]
    std::vector<int> array1 = {4, 3, 2, 1};
```

File: sorting/cycle_sort.cpp [hunk 3]
```diff
@@ -98,6 +96,20 @@ static void test() {
     std::vector<double> arr2 = sorting::cycle_sort::cycleSort(array2);
     assert(std::is_sorted(std::begin(arr2), std::end(arr2)));
     std::cout << "passed" << std::endl;
+
+    // [3, 3, 3, 3] return [3, 3, 3, 3]
+    std::vector<int> array3 = {3, 3, 3, 3};
```
Comment (mishraabhinn) [inline sorting/cycle_sort.cpp:101]:
```suggestion
     // Test 3
    // [3, 3, 3, 3] return [3, 3, 3, 3]
    std::vector<int> array3 = {3, 3, 3, 3};
```

File: sorting/cycle_sort.cpp (hunk 2 above)
Comment (mishraabhinn) [inline sorting/cycle_sort.cpp:87]:
```suggestion
     // Test 2
    // [4, 3, 2, 1] return [1, 2, 3, 4]
    std::vector<int> array1 = {4, 3, 2, 1};
```

File: sorting/cycle_sort.cpp [hunk 4]
```diff
@@ -98,6 +96,20 @@ static void test() {
...
+    std::cout << "Test 3... ";
+    std::vector<int> arr3 = sorting::cycle_sort::cycleSort(array3);
+    assert(std::is_sorted(std::begin(arr3), std::end(arr3)));
+    std::cout << "passed" << std::endl;
+
+    // [9, 4, 6, 8, 14, 3] return [9, 4, 6, 8, 14, 3]
+    std::vector<int> array4 = {3, 4, 6, 8, 9, 14};
```
Comment (mishraabhinn) [inline sorting/cycle_sort.cpp:108]:
```suggestion
     // Test 4
    // [9, 4, 6, 8, 14, 3] return [9, 4, 6, 8, 14, 3]
    std::vector<int> array4 = {3, 4, 6, 8, 9, 14};
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Comment by Swastyy** [comment]
Hey, after committing the above suggestions I got mail as Run failed: CodeQL... but here everything is working fine. Isn't it?

**Comment by mishraabhinn** [comment]
> Hey, after committing the above suggestions I got mail as Run failed: CodeQL... but here everything is working fine. Isn't it?

Is it working now?

**Comment by Swastyy** [comment]
> > Hey, after committing the above suggestions I got mail as Run failed: CodeQL... but here everything is working fine. Isn't it?
> 
> Is it working now?

Yea, I think so. 😕 
Like git bot has pushed it so there aren't any issues.

**Comment by mishraabhinn** [comment]
> > > Hey, after committing the above suggestions I got mail as Run failed: CodeQL... but here everything is working fine. Isn't it?
> > 
> > 
> > Is it working now?
> 
> Yea, I think so. 😕
> Like git bot has pushed it so there aren't any issues.

If you got mail  there is issue. Something went wrong. Go to workflow and see the error.

File: sorting/cycle_sort.cpp [hunk 5]
```diff
@@ -98,9 +98,9 @@ static void test() {
     std::cout << "passed" << std::endl;
 
     // [4.3, -6.5, -7.4, 0, 2.7, 1.8] return [-7.4, -6.5, 0, 1.8, 2.7, 4.3]
-    std::vector<double> array2 = {4.3, -6.5, -7.4, 0, 2.7, 1.8};
+    std::vector<int> array2 = {4.3, -6.5, -7.4, 0, 2.7, 1.8};
```
Comment (mishraabhinn) [inline sorting/cycle_sort.cpp:101]: Why int ?

Comment (Swastyy) [reply sorting/cycle_sort.cpp:101]: Sorry 1 sec

Comment (Swastyy) [reply sorting/cycle_sort.cpp]: 20 minutes and no mail yet 😕

**Comment by Swastyy** [comment]
@mishraabhinn errors resolved 😃

File: sorting/cycle_sort.cpp [hunk 6]
```diff
@@ -2,14 +2,12 @@
  * @file
  * @brief Implementation of [Cycle
  * sort](https://en.wikipedia.org/wiki/Cycle_sort) algorithm
- *
  * @details
- * Cycle Sort is a sorting algorithm that works in \f$O(n^2)\f$ time in best cas
- * and works in \f$O(n^2)\f$ in worst case. If a element is already at its
+ * Cycle Sort is a sorting algorithm that works in \f$O(n^2)\f$ time in best
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:6]:
```suggestion
 * Cycle Sort is a sorting algorithm that works in \f$O(n^2)\f$ time in the best
```

File: sorting/cycle_sort.cpp [hunk 7]
```diff
@@ -98,6 +103,21 @@ static void test() {
     std::vector<double> arr2 = sorting::cycle_sort::cycleSort(array2);
     assert(std::is_sorted(std::begin(arr2), std::end(arr2)));
     std::cout << "passed" << std::endl;
+
+    // Test 3
+    // [3, 3, 3, 3] return [3, 3, 3, 3]
+    std::vector<int> array3 = {3, 3, 3, 3};
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:109]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: sorting/cycle_sort.cpp (hunk 7 above)
Comment (Swastyy) [reply sorting/cycle_sort.cpp]: Done. Pl see once.

**Review by Panquesito7 — APPROVED** [review]
Looks good. Thank you for your contribution! 👍 🎉

**Review by suprithars111 — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
Earlier approval by mistake.
```
test.cpp: In instantiation of ‘std::vector<T> sorting::cycle_sort::cycleSort(const std::vector<T>&) [with T = unsigned int]’:
test.cpp:96:64:   required from here
test.cpp:53:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   53 |         if (pos == cycle_start) {
      |             ~~~~^~~~~~~~~~~~~~
test.cpp:59:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   59 |         if (pos == cycle_start) {
      |             ~~~~^~~~~~~~~~~~~~
test.cpp:65:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   65 |         while (pos != cycle_start) {
      |                ~~~~^~~~~~~~~~~~~~
test.cpp: In instantiation of ‘std::vector<T> sorting::cycle_sort::cycleSort(const std::vector<T>&) [with T = double]’:
test.cpp:103:62:   required from here
test.cpp:53:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   53 |         if (pos == cycle_start) {
      |             ~~~~^~~~~~~~~~~~~~
test.cpp:59:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   59 |         if (pos == cycle_start) {
      |             ~~~~^~~~~~~~~~~~~~
test.cpp:65:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   65 |         while (pos != cycle_start) {
      |                ~~~~^~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

shell returned 1
```

**Comment by Swastyy** [comment]
> Earlier approval by mistake.
> 
> ```
> test.cpp: In instantiation of ‘std::vector<T> sorting::cycle_sort::cycleSort(const std::vector<T>&) [with T = unsigned int]’:
> test.cpp:96:64:   required from here
> test.cpp:53:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    53 |         if (pos == cycle_start) {
>       |             ~~~~^~~~~~~~~~~~~~
> test.cpp:59:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    59 |         if (pos == cycle_start) {
>       |             ~~~~^~~~~~~~~~~~~~
> test.cpp:65:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    65 |         while (pos != cycle_start) {
>       |                ~~~~^~~~~~~~~~~~~~
> test.cpp: In instantiation of ‘std::vector<T> sorting::cycle_sort::cycleSort(const std::vector<T>&) [with T = double]’:
> test.cpp:103:62:   required from here
> test.cpp:53:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    53 |         if (pos == cycle_start) {
>       |             ~~~~^~~~~~~~~~~~~~
> test.cpp:59:17: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    59 |         if (pos == cycle_start) {
>       |             ~~~~^~~~~~~~~~~~~~
> test.cpp:65:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
>    65 |         while (pos != cycle_start) {
>       |                ~~~~^~~~~~~~~~~~~~
> cc1plus: all warnings being treated as errors
> 
> shell returned 1
> ```

Sorry for this have changed the data type of iterators.

**Review by Panquesito7 — APPROVED** [review]
All fine for me. I'll merge this now. If anybody finds any objections, please make another PR to fix them. Thanks! 😄

**Pull request merged**